File: RNIntl/GettextParser.m
//
//  GettextParser.m
//  RNIntl
//
//  Created by Taegon Kim on 2015. 12. 12..
//  Copyright © 2015 Taegon Kim. All rights reserved.
//

#import "GettextParser.h"

#define MAGIC_NUMBER 0x950412de

typedef struct {
    uint32_t magic;
    uint32_t revision;
    uint32_t stringCount;
    uint32_t originalStringOffset;
    uint32_t translationStringOffset;
    uint32_t hashTableSize;
    uint32_t hashTableOffset;
} MoHeader;

typedef struct {
    uint32_t length;
    uint32_t offset;
} Position;

@implementation GettextParser

@synthesize lastError;
@synthesize catalog;

- (instancetype)initWithFileAtPath:(nonnull NSString *)filepath {
    if (self = [super init]) {
        [self loadFileAtPath:filepath];
    }
    return self;
}

- (BOOL)loadFileAtPath:(NSString *)filepath {
    MoHeader header = { 0 };
    NSError *err = nil;
    NSData *data = [NSData dataWithContentsOfFile:filepath options:NSDataReadingMappedIfSafe error:&err];

    if (err != nil) {
      lastError = err;
      return NO;
    }

    // get header
    [data getBytes:&header length:sizeof(header)];

    // check magic number
    if (header.magic != MAGIC_NUMBER && header.magic != OSSwapInt32(MAGIC_NUMBER)) {
      return NO;
    }

    NSDictionary *headers = nil;
    NSMutableDictionary *messages = [[NSMutableDictionary alloc] init];
    for (uint i = 0; i < header.stringCount; i++) {
        Position pos = { 0 };
        uint32_t offset = 0;

        // msgid - original string
        offset = header.originalStringOffset + i * sizeof(pos);
        [data getBytes:&pos range:NSMakeRange(offset, sizeof(pos))];
        NSArray *idList = [self getStringsFromData:[data subdataWithRange:NSMakeRange(pos.offset, pos.length)]];

        // msgstr - translation string
        offset = header.translationStringOffset + i * sizeof(pos);
        [data getBytes:&pos range:NSMakeRange(offset, sizeof(pos))];
        NSArray *translationList = [self getStringsFromData:[data subdataWithRange:NSMakeRange(pos.offset, pos.length)]];

        for (uint j = 0; j < idList.count; j++) {
            if ([[idList objectAtIndex:0] isEqualToString:@""]) {
                headers = [self parseHeader:[translationList objectAtIndex:0]];
            } else {
                [messages setValue:translationList forKey:[idList objectAtIndex:j]];
            }
        }
    }

    catalog = @{ @"headers": headers, @"translations": messages };

    return YES;
}

- (NSDictionary *)parseHeader:(nonnull NSString *)headerString {
    NSArray *_headers = [headerString componentsSeparatedByString:@"\n"];
    NSMutableDictionary *headers = [[NSMutableDictionary alloc] init];

    for (int i=0; i < _headers.count; i++) {
        NSRange range = [_headers[i] rangeOfString:@":"];
        if (range.location == NSNotFound) continue;

        NSString *key = [_headers[i] substringToIndex:range.location];
        NSString *value = [_headers[i] substringFromIndex:range.location+range.length];

        key = [key stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        value = [value stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];

        [headers setValue:value forKey:key];
    }

    return headers;
}

- (NSDictionary *)getCatalog {
    return catalog;
}

- (NSArray *)getStringsFromData:(nonnull NSData *) data{
    NSMutableArray *result = [[NSMutableArray alloc] init];
    NSData *pattern = [[NSData alloc] initWithBytes:(unsigned char[]){0x04} length:1];
    NSRange range = [data rangeOfData:pattern options:0 range:NSMakeRange(0, data.length)];
    NSUInteger from = 0;

    // strip context
    if (range.location != NSNotFound) {
        data = [data subdataWithRange:NSMakeRange(0, range.location)];
    }

    // null pattern
    pattern = [[NSData alloc] initWithBytes:(unsigned char[]){0x00} length:1];

    do {
        range = [data rangeOfData:pattern options:0 range:NSMakeRange(from, data.length - from)];
        if (range.location == NSNotFound) break;

        NSString *str = [[NSString alloc] initWithData:[data subdataWithRange:NSMakeRange(from, range.location - from)] encoding:NSUTF8StringEncoding];
        [result addObject:str];

        from = range.location + range.length;
    } while (TRUE);

    if (result.count == 0 && data.length == 0) {
        [result addObject:@""];
    } else {
        if (from > 0) {
            data = [data subdataWithRange:NSMakeRange(from, data.length - from)];
        }
        [result addObject:[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]];
    }

    return result;
}

@end

File: android/src/main/java/kim/taegon/rnintl/GettextParser.java
/**
 * Created by taggon
 */
package kim.taegon.rnintl;

import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.HashMap;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.nio.ByteBuffer;

public class GettextParser {
    public final static long MAGIC_NUMBER = 0x950412DEL;
    public final static long LE_MAGIC_NUMBER = 0xDE120495L;
    public final static long UINT = 0xffffffffL;

    protected String lastError = "";
    protected Map<String, Object> catalog = new HashMap<>();
    protected boolean isLittleEndian = false;

    public GettextParser(String filepath) {
        loadFile(filepath);
    }

    public GettextParser(InputStream stream) {
        loadFile(stream);
    }

    public boolean loadFile(String filepath) {
        try {
            InputStream stream = new FileInputStream(filepath);
            loadFile(stream);
            stream.close();
        } catch (Exception e) {
            lastError = e.getMessage();
            return false;
        }

        return true;
    }

    public boolean loadFile(InputStream stream) {
        Map<String, String> headers = new HashMap<>();
        Map<String, String[]> translations = new HashMap<>();

        try {
            DataInputStream ds = new DataInputStream(stream);

            Map<String, Integer> fileHeader = readFileHeader(ds);

            for (int i = 0; i < fileHeader.get("stringCount"); i++) {
                // msgid - original string
                String[] idList = getStringsFromFile(ds, fileHeader.get("originalStringOffset") + i * 8);

                // msgstr - translation string
                String[] translationList = getStringsFromFile(ds, fileHeader.get("translationStringOffset") + i * 8);

                if (idList.length == 1 && idList[0].isEmpty()) {
                    headers = parseHeader(translationList[0]);
                    continue;
                }

                for (String msgid: idList) {
                    translations.put(msgid, translationList);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            lastError = e.getMessage();
            return false;
        }

        catalog.put("headers", headers);
        catalog.put("translations", translations);

        return true;
    }

    protected Map<String, Integer> readFileHeader(DataInputStream ds) throws Exception {
        Map<String, Integer> header = new HashMap<>();
        long magicNumber = 0;

        try {
            // reset the position
            ds.reset();

            // magic number
            magicNumber = ds.readInt() & UINT;

            if (magicNumber != MAGIC_NUMBER && magicNumber != LE_MAGIC_NUMBER) {
                throw new Exception("Invalid magic number");
            }

            // little endian?
            isLittleEndian = (magicNumber == LE_MAGIC_NUMBER);

            // revision
            header.put("revision", swapInteger(ds.readInt()));

            // string count
            header.put("stringCount", swapInteger(ds.readInt()));

            // original string offset
            header.put("originalStringOffset", swapInteger(ds.readInt()));

            // translation string offset
            header.put("translationStringOffset", swapInteger(ds.readInt()));

            // hash table size
            header.put("hashTableSize", swapInteger(ds.readInt()));

            // hash table offset
            header.put("hashTableOffset", swapInteger(ds.readInt()));
        } catch(Exception e) {
            throw e;
        }

        return header;
    }

    final protected int swapInteger(int n) {
        if (isLittleEndian) {
            byte[] b = ByteBuffer.allocate(4).putInt(n).array();
            n = ((b[3]&0xff)<<24) + ((b[2]&0xff)<<16) + ((b[1]&0xff)<<8) + (b[0]&0xff);
        }
        return n;
    }

    protected String[] getStringsFromFile(DataInputStream ds, int offset) {
        List<String> result = new ArrayList<>();
        int len = 0, pos = 0;

        try {
            ds.reset();
            ds.skip(offset);
            len = swapInteger(ds.readInt());
            pos = swapInteger(ds.readInt());

            byte[] data = new byte[len];
            ds.reset();
            ds.skip(pos);
            ds.read(data);

            // strip context
            int contextSeparatorPos = -1;
            for (int i = 0; i < data.length; i++) {
                if (data[i] == 4) {
                    contextSeparatorPos = i;
                    break;
                }
            }
            if (contextSeparatorPos > -1) {
                len -= contextSeparatorPos;
                data = Arrays.copyOfRange(data, contextSeparatorPos+1, len);
            }

            int nullSeparatorPos = 0;
            do {
                nullSeparatorPos = -1;
                for (int i = 0; i < data.length; i++) {
                    if (data[i] == 0) {
                        nullSeparatorPos = i;
                        break;
                    }
                }
                if (nullSeparatorPos < 0) break;

                String str = new String(Arrays.copyOfRange(data, 0, nullSeparatorPos), StandardCharsets.UTF_8);
                result.add(str);

                data = Arrays.copyOfRange(data, nullSeparatorPos+1, data.length);
            } while(true);

            if (result.size() == 0 && data.length == 0) {
                result.add("");
            } else if (data.length > 0) {
                String str = new String(data, StandardCharsets.UTF_8);
                result.add(str);
            }
        } catch (Exception e) {
            lastError = e.getMessage();
            return null;
        }

        return result.toArray(new String[result.size()]);
    }

    public Map<String, String> parseHeader(String headerString) {
        String[] headers = headerString.split("\\n");
        Map<String, String> result = new HashMap<>();

        for(String header: headers) {
            String[] entry = header.split("\\s*:\\s*", 2);
            if (entry.length < 2) continue;

            result.put(entry[0], entry[1]);
        }

        return result;
    }

    public Map<String, Object> getCatalog() {
        return catalog;
    }
}
